src/hist.jl

```diff
@@ -140,100 +155,270 @@ function show(io::IO, h::AbstractHistogram)
+fit{T,N,W}(::Type{Histogram{T}}, vs::NTuple{N,AbstractVector}, wv::WeightVec{W}, edges::NTuple{N,AbstractVector}; closed::Symbol=:default_left) =
+    append!(Histogram(edges, T, _check_closed_arg(closed,:fit), false), vs, wv)
+
+fit{T,N}(::Type{Histogram{T}}, vs::NTuple{N,AbstractVector}, wv::WeightVec; closed::Symbol=:default_left, isdensity::Bool=false, nbins=sturges(length(vs[1]))) = begin
     closed = _check_closed_arg(closed,:fit)
-    fit(Histogram, v, wv, histrange(v,nbins,closed); closed=closed)
+    fit(Histogram{T}, vs, wv, histrange(vs,nbins,closed); closed=closed)
 end
 
-# N-dimensional
-function push!{T,N}(h::Histogram{T,N},xs::NTuple{N,Real},w::Real)
-    is = if h.closed == :right
-        map((edge, x) -> searchsortedfirst(edge,x) - 1, h.edges, xs)
-    else
-        map(searchsortedlast, h.edges, xs)
-    end
-    try
-        h.weights[is...] += w
-    catch e
-        isa(e,BoundsError) || rethrow(e)
+fit(::Type{Histogram}, args...; kwargs...) = fit(Histogram{Int}, args...; kwargs...)
+fit{N,W}(::Type{Histogram}, vs::NTuple{N,AbstractVector}, wv::WeightVec{W}, args...; kwargs...) = fit(Histogram{W}, vs, wv, args...; kwargs...)
+
+
+# Get a suitable high-precision type for the norm of a histogram.
+@generated function norm_type{T, N, E}(h::Histogram{T, N, E})
+    args = [:( eltype(edges[$d]) ) for d = 1:N]
+    quote
+        edges = h.edges
+        norm_type(promote_type(T, $(args...)))
     end
-    h
 end
-push!{T,N}(h::AbstractHistogram{T,N},xs::NTuple{N,Real}) = push!(h,xs,one(T))
 
-function append!{T,N}(h::AbstractHistogram{T,N}, vs::NTuple{N,AbstractVector})
-    for xs in zip(vs...)
-        push!(h,xs)
+norm_type{T<:Integer}(::Type{T}) = promote_type(T, Int64)
+norm_type{T<:AbstractFloat}(::Type{T}) = promote_type(T, Float64)
+
+
+"""
+    norm(h::Histogram)
+
+Calculate the norm of histogram `h` as the absolute value of its integral.
+"""
+@generated function norm{T, N, E}(h::Histogram{T, N, E})
+    quote
+        edges = h.edges
+        weights = h.weights
+        SumT = norm_type(h)
+        v_0 = 1
+        s_0 = zero(SumT)
+        @inbounds @nloops(
+            $N, i, weights,
+            d -> begin
+                v_{$N-d+1} = v_{$N-d} * _edge_binvolume(SumT, edges[d], i_d)
+                s_{$N-d+1} = zero(SumT)
+            end,
+            d -> begin
+                s_{$N-d} += s_{$N-d+1}
+            end,
+            begin
+                $(Symbol("s_$(N)")) += (@nref $N weights i) * $(Symbol("v_$N"))
+            end
+        )
+        s_0
     end
-    h
 end
-function append!{T,N}(h::AbstractHistogram{T,N}, vs::NTuple{N,AbstractVector},wv::WeightVec)
-    for (xs,w) in zip(zip(vs...),wv.values)
-        push!(h,xs,w)
+
+
+function _float_deepcopy(x)
+    y = float(x)
+    y === x? deepcopy(y) : y
+end
+
+
+function float{T, N, E}(h::Histogram{T, N, E})
+    float_weights = float(h.weights)
+    if float_weights === h.weights
```

Comment:
> @andreasnoack: My point here is that instead of the branch you can write two methods

I understand - I just forgot to delete the other version, somehow, sorry. Fixed now.